File: common/src/main/java/com/wolfyscript/utilities/gui/ViewRuntimeImpl.java
package com.wolfyscript.utilities.gui;

import com.wolfyscript.utilities.WolfyUtils;
import com.wolfyscript.utilities.gui.callback.TextInputCallback;
import com.wolfyscript.utilities.gui.callback.TextInputTabCompleteCallback;
import com.wolfyscript.utilities.gui.interaction.InteractionHandler;
import com.wolfyscript.utilities.gui.model.UpdateInformation;
import com.wolfyscript.utilities.gui.reactivity.ReactiveSourceImpl;
import com.wolfyscript.utilities.gui.rendering.Renderer;
import com.wolfyscript.utilities.gui.rendering.RenderingGraph;

import java.util.*;
import java.util.function.Function;

public class ViewRuntimeImpl implements ViewRuntime {

    private static long NEXT_ID = Long.MIN_VALUE;

    private final long id;
    private final RenderingGraph renderingGraph;
    private final Renderer<?> renderer;
    private final InteractionHandler interactionHandler;

    private final WolfyUtils wolfyUtils;
    private final Router router;
    private Window currentRoot;
    private final Deque<Window> history;
    private final Set<UUID> viewers;

    private TextInputCallback textInputCallback;
    private TextInputTabCompleteCallback textInputTabCompleteCallback;

    private final ReactiveSourceImpl reactiveSource;

    protected ViewRuntimeImpl(WolfyUtils wolfyUtils, Function<ViewRuntime, RouterBuilder> rootRouter, Set<UUID> viewers) {
        this.wolfyUtils = wolfyUtils;
        this.renderingGraph = new RenderingGraph(this);
        this.reactiveSource = new ReactiveSourceImpl(this);
        this.renderer = wolfyUtils.getCore().platform().guiUtils().createRenderer(this);
        this.interactionHandler = wolfyUtils.getCore().platform().guiUtils().createInteractionHandler(this);
        this.router = rootRouter.apply(this).create(null);

        this.history = new ArrayDeque<>();
        this.viewers = viewers;
        id = NEXT_ID++;
    }

    public InteractionHandler getInteractionHandler() {
        return interactionHandler;
    }

    public RenderingGraph getRenderingGraph() {
        return renderingGraph;
    }

    public ReactiveSourceImpl getReactiveSource() {
        return reactiveSource;
    }

    public void incomingUpdate(UpdateInformation information) {
        interactionHandler.update(information);
        renderer.update(information);
    }

    @Override
    public void openNew() {
        openNew(new String[0]);
    }

    @Override
    public void open() {
        if (history.isEmpty()) {
            openNew();
        } else {
            getCurrentMenu().ifPresent(window -> {
                renderer.changeWindow(window);
                interactionHandler.init(window);
                window.open(this);
                setCurrentRoot(window);
            });
        }
    }

    @Override
    public void openPrevious() {
        history.poll(); // Remove active current menu
        setCurrentRoot(history.peek());
    }

    public void setCurrentRoot(Window currentRoot) {
        getCurrentMenu().ifPresent(window -> window.close(this));
        this.currentRoot = currentRoot;
    }

    @Override
    public Optional<Window> getCurrentMenu() {
        return Optional.ofNullable(currentRoot);
    }

    @Override
    public WolfyUtils getWolfyUtils() {
        return wolfyUtils;
    }

    @Override
    public Router getRouter() {
        return router;
    }

    @Override
    public Set<UUID> getViewers() {
        return Set.copyOf(viewers);
    }

    public long getId() {
        return id;
    }

    @Override
    public Optional<TextInputCallback> textInputCallback() {
        return Optional.ofNullable(textInputCallback);
    }

    @Override
    public void setTextInputCallback(TextInputCallback textInputCallback) {
        this.textInputCallback = textInputCallback;
    }

    @Override
    public Optional<TextInputTabCompleteCallback> textInputTabCompleteCallback() {
        return Optional.ofNullable(textInputTabCompleteCallback);
    }

    @Override
    public void setTextInputTabCompleteCallback(TextInputTabCompleteCallback textInputTabCompleteCallback) {
        this.textInputTabCompleteCallback = textInputTabCompleteCallback;
    }

    @Override
    public long id() {
        return id;
    }

    @Override
    public void openNew(String... path) {
        Window window = getRouter().open(this, path);
        setCurrentRoot(window);
        renderer.changeWindow(window);
        interactionHandler.init(window);
        wolfyUtils.getCore().platform().scheduler().syncTask(wolfyUtils, renderer::render);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ViewRuntimeImpl that = (ViewRuntimeImpl) o;
        return id == that.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: common/src/main/java/com/wolfyscript/utilities/gui/ViewRuntimeImpl.kt
package com.wolfyscript.utilities.gui

import com.wolfyscript.utilities.WolfyUtils
import com.wolfyscript.utilities.gui.interaction.InteractionHandler
import com.wolfyscript.utilities.gui.model.UpdateInformation
import com.wolfyscript.utilities.gui.reactivity.ReactiveSourceImpl
import com.wolfyscript.utilities.gui.rendering.Renderer
import com.wolfyscript.utilities.gui.rendering.ModelGraph
import java.util.*
import java.util.function.Function

class ViewRuntimeImpl(
    override val wolfyUtils: WolfyUtils,
    rootRouter: Function<ViewRuntime, Window>,
    override val viewers: Set<UUID>,
) : ViewRuntime {
    @JvmField
    val id: Long = NEXT_ID++

    // Create rendering & reactivity trees
    val modelGraph: ModelGraph = ModelGraph(this)
    val reactiveSource: ReactiveSourceImpl = ReactiveSourceImpl(this)

    // Create platform specific handlers that handle rendering and interaction
    val renderer: Renderer<*> = wolfyUtils.core.platform.guiUtils.createRenderer(this)

    val interactionHandler: InteractionHandler = wolfyUtils.core.platform.guiUtils.createInteractionHandler(this)
    // Build the components and init the rendering tree
    private var currentRoot: Window? = rootRouter.apply(this)
    override val currentMenu: Window?
        get() = currentRoot

    private val history: Deque<Window> = ArrayDeque()

    fun incomingUpdate(information: UpdateInformation?) {
        interactionHandler.update(information!!)
        renderer.update(information)
    }

    override fun openNew() {
        openNew(*emptyArray())
    }

    override fun openNew(vararg path: String) {
        currentMenu?.apply { open(this) }
    }

    override fun open() {
        if (history.isEmpty()) {
            currentMenu?.close()
            openNew()
        } else {
            currentMenu?.let { open(it) }
        }
    }

    private fun open(window: Window) {
        setCurrentRoot(window)

        renderer.changeWindow(window)
        interactionHandler.init(window)

        wolfyUtils.core.platform.scheduler.syncTask(wolfyUtils) {
            renderer.render()
            reactiveSource.owner()?.update()
        }
    }

    override fun openPrevious() {
        history.poll() // Remove active current menu
        val window = history.peek()
        currentMenu?.close()
        open(window)
    }

    fun setCurrentRoot(currentRoot: Window?) {
        this.currentRoot = currentRoot
    }

    fun getCurrentMenu(): Optional<Window> {
        return Optional.ofNullable(currentRoot)
    }

    override fun id(): Long {
        return id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as ViewRuntimeImpl
        return id == that.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    companion object {
        private var NEXT_ID = Long.MIN_VALUE
    }
}
